pkg/gator/expand.go

```diff
@@ -0,0 +1,225 @@
+package gator
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/apis/expansion/unversioned"
+	mutationsunversioned "github.com/open-policy-agent/gatekeeper/apis/mutations/unversioned"
+	"github.com/open-policy-agent/gatekeeper/pkg/expansion"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutators/assign"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutators/assignmeta"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/mutators/modifyset"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+var (
+	MutatorKinds = map[string]bool{
+		"Assign":         true,
+		"AssignMetadata": true,
+		"ModifySet":      true,
+	}
+	MutatorAPIVersions = map[string]bool{
+		"mutations.gatekeeper.sh/v1alpha1": true,
+		"mutations.gatekeeper.sh/v1beta1":  true,
+	}
+	ExpansionAPIVersions = map[string]bool{
+		"expansion.gatekeeper.sh/v1alpha1": true,
+	}
+)
+
+type expansionResources struct {
+	mutators           []types.Mutator
+	templateExpansions []*unversioned.TemplateExpansion
+	generators         []*unstructured.Unstructured
+	namespaces         map[string]*corev1.Namespace
+}
+
+func Expand(resources []*unstructured.Unstructured) ([]*unstructured.Unstructured, error) {
+	mutSystem := mutation.NewSystem(mutation.SystemOpts{})
+	expSystem := expansion.NewSystem()
+	er := expansionResources{}
+	if err := er.addResources(resources); err != nil {
+		return nil, fmt.Errorf("error parsing resources: %s", err)
+	}
+
+	for _, te := range er.templateExpansions {
+		if err := expSystem.UpsertTemplate(te); err != nil {
+			return nil, err
+		}
+	}
+
+	var resultants []*unstructured.Unstructured
+	for _, gen := range er.generators {
+		ns, err := er.namespaceForGenerator(gen)
+		base := &types.Mutable{
+			Object:    gen,
+			Namespace: ns,
+			Username:  "",
+			Source:    types.SourceTypeGenerated,
+		}
+		if err != nil {
+			return nil, fmt.Errorf("error expanding generator: %s", err)
+		}
+		r, err := expSystem.Expand(base, mutSystem)
+		if err != nil {
+			return nil, fmt.Errorf("error expanding generator: %s", err)
+		}
+		resultants = append(resultants, r...)
+	}
+
+	return resultants, nil
+}
+
+func (er *expansionResources) namespaceForGenerator(gen *unstructured.Unstructured) (*corev1.Namespace, error) {
+	genNs := gen.GetNamespace()
+	if genNs == "" {
+		return &corev1.Namespace{}, nil
+	}
+	if genNs == "default" {
```

Comment:
if `default` happens to be defined in the source, we should retrieve it from `er.namespaces`